src/Subscriber/PaymentSubscriber.php

```diff
@@ -199,9 +208,32 @@ public static function getSubscribedEvents(): array
      */
     public function onContextTokenUpdate(SalesChannelContextSwitchEvent $event)
     {
-        // Save (or clear) state.data if payment method is selected/updated
+        // Clear state.data if payment method is updated
         if ($event->getRequestDataBag()->has('paymentMethodId')) {
-            $this->saveStateData($event);
+            $this->removeCurrentStateData($event);
+        }
+
+        // Save state data, only if Adyen payment method is selected
+        if ($event->getRequestDataBag()->get('adyenStateData')) {
+            // Use payment method selected in the same request if available, otherwise get payment method from context
+            $paymentMethodId = $event->getRequestDataBag()->get('paymentMethodId')
+                ?? $event->getSalesChannelContext()->getPaymentMethod()->getId();
+            if (!$paymentMethodId) {
+                $this->logger->error('A payment method must be selected before saving payment state data.');
+                $this->session->getFlashBag()
```

Comment:
btw. not really relevant but could you maybe use the validation scheme like the ContextSwitchroute does because upon using the store API, the flash messages in the session are completely irrelevant 😄